File: week-4/problem-set/prob2.py
"""Prob2 example.

This file belongs to the week 4 divide-and-conquer and memoization examples. It uses find_min_max to illustrate the main idea with a small runnable example.
"""

def find_min_max(arr, low, high):
    """Return the minimum and maximum values using divide and conquer.

    Args:
        arr (list): The list being processed by the algorithm.
        low (int): The low input used by this function.
        high (int): The high input used by this function.

    Returns:
        tuple: The tuple of values produced by the algorithm.
    """
    if low == high:
        return (arr[low], arr[low])
    elif high == low + 1:
        return (min(arr[low], arr[high]), max(arr[low], arr[high]))
    else:
        mid = (low + high) // 2
        (min1, max1) = find_min_max(arr, low, mid)
        (min2, max2) = find_min_max(arr, mid+1, high)
        return (min(min1, min2), max(max1, max2))

# Example usage
arr = [3, 7, 1, 8, 2, 5]
n = len(arr)
(minimum, maximum) = find_min_max(arr, 0, n-1)

print("Array:", arr)
print("Minimum:", minimum)
print("Maximum:", maximum)

File: week-4/problem-set/prob2.cpp
/**
 * @file prob2.cpp
 * @brief C++ version of the divide-and-conquer minimum/maximum example.
 */

#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

/**
 * @brief Return the minimum and maximum values in a range using divide and conquer.
 * @param arr const vector<int>& Input list of integers.
 * @param low int Left boundary of the current recursive range.
 * @param high int Right boundary of the current recursive range.
 * @return pair<int, int> Pair containing (minimum, maximum) for arr[low..high].
 */
pair<int, int> find_min_max(const vector<int>& arr, int low, int high) {
    if (low == high) {
        return {arr[low], arr[low]};
    }
    if (high == low + 1) {
        return {min(arr[low], arr[high]), max(arr[low], arr[high])};
    }

    int mid = (low + high) / 2;
    pair<int, int> left = find_min_max(arr, low, mid);
    pair<int, int> right = find_min_max(arr, mid + 1, high);
    return {min(left.first, right.first), max(left.second, right.second)};
}

/**
 * @brief Run the minimum/maximum demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> arr = {3, 7, 1, 8, 2, 5};
    pair<int, int> result = find_min_max(arr, 0, static_cast<int>(arr.size()) - 1);
    cout << "Array: ";
    for (int value : arr) {
        cout << value << ' ';
    }
    cout << "\nMinimum: " << result.first << "\nMaximum: " << result.second << '\n';
    return 0;
}
